airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonSchemaToAirbyteType.kt

```diff
@@ -107,60 +110,48 @@ class JsonSchemaToAirbyteType {
         if (properties.isEmpty) {
             return ObjectTypeWithEmptySchema
         }
-        val requiredFields = schema.get("required")?.map { it.asText() } ?: emptyList()
-        return objectFromProperties(properties as ObjectNode, requiredFields)
+        val propertiesMapped =
+            properties
+                .fields()
+                .asSequence()
+                .map { (name, node) -> name to fieldFromSchema(node as ObjectNode) }
+                .toMap(LinkedHashMap())
+        return ObjectType(propertiesMapped)
     }
 
     private fun fieldFromSchema(
         fieldSchema: ObjectNode,
-        onRequiredList: Boolean = false
     ): FieldType {
-        val markedRequired = fieldSchema.get("required")?.asBoolean() ?: false
-        val nullable = !(onRequiredList || markedRequired)
-        val airbyteType = convert(fieldSchema)
-        if (airbyteType is UnionType) {
-            return fieldFromUnion(airbyteType, nullable)
-        } else {
-            return FieldType(airbyteType, nullable)
-        }
-    }
-
-    private fun fieldFromUnion(unionType: UnionType, nullable: Boolean = false): FieldType {
-        if (unionType.options.contains(NullType)) {
-            val filtered = unionType.options.filter { it != NullType }
-            return FieldType(UnionType(filtered), nullable = true)
-        }
-        return FieldType(unionType, nullable = nullable)
-    }
-
-    private fun objectFromProperties(schema: ObjectNode, requiredFields: List<String>): ObjectType {
-        val properties =
-            schema
-                .fields()
-                .asSequence()
-                .map { (name, node) ->
-                    name to fieldFromSchema(node as ObjectNode, requiredFields.contains(name))
-                }
-                .toMap(LinkedHashMap())
-        return ObjectType(properties)
+        val airbyteType =
+            convertInner(fieldSchema) ?: UnknownType("Illegal null type as field type")
+        return FieldType(airbyteType, nullable = true)
     }
 
     private fun unionFromCombinedTypes(
         options: List<JsonNode>,
         parentSchema: ObjectNode
-    ): UnionType {
+    ): AirbyteType {
         // Denormalize the properties across each type (the converter only checks what matters
         // per type).
         val unionOptions =
-            options.map {
+            options.mapNotNull {
                 if (it.isTextual) {
                     val schema = parentSchema.deepCopy()
                     schema.put("type", it.textValue())
-                    convert(schema)
+                    convertInner(schema)
                 } else {
-                    convert(it)
+                    convertInner(it)
                 }
             }
-        return UnionType(unionOptions)
+        return optionsToUnionOrSingleType(unionOptions)
     }
+
+    private fun optionsToUnionOrSingleType(options: List<AirbyteType>): AirbyteType =
```

Comment:
Pursuant to early comment, basically move this into a companion object in the union type, and block direct construction